crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+    fn query<'a, I>(
+        circuit: &mut DBSPHandle,
+        hqueries: &ZSetHandle<SqlString>,
+        houtput_strings: &OutputHandle<OrdZSet<SqlString>>,
+        queries: I,
+    ) where
+        I: IntoIterator<Item = &'a str>,
+    {
+        let mut queries = queries.into_iter().map(SqlString::from).collect::<Vec<_>>();
+
+        let mut tuples = queries
+            .iter()
+            .map(|s| Tup2(s.clone(), 1))
+            .collect::<Vec<_>>();
+        hqueries.append(&mut tuples);
+
+        circuit.step().unwrap();
+        let output = houtput_strings.consolidate();
+        let mut output = output.iter().map(|(s, _, _)| s.clone()).collect::<Vec<_>>();
+        output.sort();
+
+        queries.sort();
+        assert_eq!(output, queries);
+
+        let mut tuples = queries
+            .iter()
+            .map(|s| Tup2(s.clone(), -1))
+            .collect::<Vec<_>>();
+        hqueries.append(&mut tuples);
+
+        circuit.step().unwrap();
+    }
+
+    #[test]
+    #[serial]
+    fn test_interner_basic() {
+        let path = tempfile::tempdir().unwrap().keep();
+
+        let (mut circuit, (hinput_strings, hqueries, houtput_strings)) =
+            interner_test_circuit(path.as_path(), None);
+
+        hinput_strings.push(SqlString::from("1"), 1);
+        query(&mut circuit, &hqueries, &houtput_strings, ["1"]);
+
+        hinput_strings.push(SqlString::from("2"), 1);
+        query(&mut circuit, &hqueries, &houtput_strings, ["2"]);
+
+        hinput_strings.push(SqlString::from("3"), 1);
+        query(&mut circuit, &hqueries, &houtput_strings, ["3"]);
+
+        hinput_strings.push(SqlString::from("4"), 1);
+        query(&mut circuit, &hqueries, &houtput_strings, ["4"]);
+
+        hinput_strings.push(SqlString::from("5"), 1);
+        query(&mut circuit, &hqueries, &houtput_strings, ["5"]);
+
+        query(
+            &mut circuit,
+            &hqueries,
+            &houtput_strings,
+            ["1", "2", "3", "4", "5"],
+        );
+
+        let checkpoint = circuit.commit().unwrap();
+        circuit.kill().unwrap();
+        super::INTERNED_STRING_CACHE.clear();
+
+        let (mut circuit, (hinput_strings, hqueries, houtput_strings)) =
+            interner_test_circuit(path.as_path(), Some(checkpoint.uuid));
+
+        query(
+            &mut circuit,
+            &hqueries,
+            &houtput_strings,
+            ["1", "2", "3", "4", "5"],
+        );
+
+        hinput_strings.push(SqlString::from("6"), 1);
+        hinput_strings.push(SqlString::from("7"), 1);
+
+        query(
+            &mut circuit,
+            &hqueries,
+            &houtput_strings,
+            ["1", "2", "3", "4", "5", "6", "7"],
+        );
+        circuit.kill().unwrap();
+    }
+
+    /// Intern a small number (1,000) of strings repeatedly.
+    #[test]
+    #[serial]
+    fn test_interner_small() {
+        // INTERNED_STRING_CACHE is a global variable, so we need to reset it
```

Comment:
how will this work when I try to run multiple tests in the same executable?
It looks to me that the circuit should be responsible for this on startup

Reply:
I serialize tests at the moment, but I might be able to do better and have a cache per `Runtime`. I'll do that.